Shorter options help

Gedochao opened the pull request:
Fixes #1669 

- bump `scala-cli-signing` and `case-app`
- only show `tags.important` or `tags.must` help options in the short help (`sub-command --help`)
- example output:
```bash
$ scala-cli fmt -h
Usage: scala-cli fmt [options]
Formats Scala code.

You are currently viewing the basic help for the fmt sub-command. You can view the full help by running: 
   scala-cli fmt --help-full
For detailed documentation refer to our website: https://scala-cli.virtuslab.org/docs/commands/fmt

Format options:
  --check                                     Check if sources are well formatted
  --save-scalafmt-conf                        Saves .scalafmt.conf file if it was created or overwritten
  -F, --scalafmt-arg string*                  Pass an argument to scalafmt.
  --scalafmt-conf, --scalafmt-config string?  Custom path to the scalafmt configuration file.
  --dialect, --scalafmt-dialect string?       Pass a global dialect for scalafmt. This overrides whatever value is configured in the .scalafmt.conf file or inferred based on Scala version used.
  --fmt-version, --scalafmt-version string?   Pass scalafmt version before running it (3.6.1 by default). If passed, this overrides whatever value is configured in the .scalafmt.conf file.

Help options:
  --help-js                                      Show options for ScalaJS
  --help-native                                  Show options for ScalaNative
  --help-doc, --doc-help, --help-scaladoc, --scaladoc-help  Show options for Scaladoc
  --help-repl, --repl-help                       Show options for Scala REPL
  --help-fmt, --fmt-help, --help-scalafmt, --scalafmt-help  Show options for Scalafmt
```

Gedochao committed:
NIT Refactor SpecificationLevel

Gedochao committed:
Limit short `--help` options to just `MUST` & `IMPORTANT` tags

Gedochao committed:
Bump `case-app` to `2.1.0-M24` & `scala-cli-signing` to `0.1.16`

Gedochao committed:
Shorten the basic help message for the `run` sub-command

Gedochao committed:
Shorten the basic help message for the `version` sub-command

Gedochao committed:
Tweak the help output for the `update` sub-command

Gedochao committed:
Tweak the help output for the `uninstall completions` sub-command

Gedochao committed:
Tweak the help output for the `test` sub-command
  - shorten the help message
  - tweak which options are included in the short help
  - misc tweaks

Gedochao committed:
Tweak the help output for the `shebang` sub-command
  - shorten the help message
  - tweak which options are included in the short help
  - misc tweaks

Gedochao committed:
Tweak the help output for the `setup-ide` sub-command
  - shorten the help message
  - tweak which options are included in the short help
  - misc tweaks

Gedochao committed:
Tweak the help output for the `repl` sub-command
  - shorten the help message
  - tweak which options are included in the short help
  - misc tweaks

Gedochao committed:
Tweak the help output for the `publish` sub-command
  - shorten the help message
  - tweak which options are included in the short help
  - misc tweaks

Gedochao committed:
Tweak the help output for the `publish setup` sub-command
  - shorten the help message
  - tweak which options are included in the short help
  - misc tweaks

Gedochao committed:
Tweak the help output for the `publish local` sub-command
  - shorten the help message
  - tweak which options are included in the short help
  - misc tweaks

Gedochao committed:
Hide non-publishing-specific options in the short help output for `publish`, `publish local` and `publish setup` sub-commands

Gedochao committed:
Tweak the short help output for the `package` sub-command

Gedochao committed:
Tweak the short help output for the `install-home` sub-command

Gedochao committed:
Tweak the short help output for the `install completions` sub-command

Gedochao committed:
Tweak the short help output for the `github secret list` & `github secret create` sub-commands

Gedochao committed:
Tweak the short help output for the `fmt` sub-command

Gedochao committed:
Tweak the short help output for the `export` sub-command

Gedochao committed:
Tweak the short help output for the `doc` sub-command

Gedochao committed:
Tweak the short help output for the `dependency-update` sub-command

Gedochao committed:
Tweak the short help output for the `config` sub-command

Gedochao committed:
Tweak the short help output for the `compile` sub-command

Gedochao committed:
Tweak the short help output for the `clean` sub-command

Gedochao committed:
Tweak the short help output for the `bsp` sub-command

Gedochao committed:
Tweak the short help output for the `bloop` sub-command

Gedochao committed:
Put the `Help` help group at the end of sorted groups

lwronski commented on modules/specification-level/src/main/scala/scala/cli/commands/SpecificationLevel.scala:
```diff
@@ -65,6 +65,9 @@ object tags {
   val should: String         = SpecificationLevel.SHOULD.toString
 
+  // other tags
+  val important: String = "important"
```
Do we need a new tag? if so, it should be well described.

lwronski replied:
And what is the difference between `Hidden` annotation and tag implementation?

 If we add `Hidden` annotation wherever there is no `implementation` tag, we will achieve the same effect.

Gedochao replied:
uh... shouldn't all options have a `SpecificationLevel` tag assigned, so that the [Runner Specification](https://scala-cli.virtuslab.org/docs/reference/scala-command/runner-specification) definition is complete?

Gedochao replied:
https://scala-cli.virtuslab.org/docs/reference/scala-command/runner-specification#implementantation-specific-options
certain options land in the `implementation specific` section there, no?

Gedochao replied:
as discussed offline, `tags.important` has been renamed to `tags.inShortHelp`

lwronski commented on website/docs/reference/commands.md:
````diff
@@ -201,25 +201,8 @@ Accepts option groups: [coursier](./cli-options.md#coursier-options), [debug](./
-```
-
+You are currently viewing the basic help for the run sub-command. You can view the full help by running: 
+   [1mscala-cli run --help-full[0m
````
It is not formatted correctly.

Gedochao replied:
It actually is formatted correctly, but not for the website...
I have mixed feelings about continuing to use console `--help` messages for the website reference docs.
`[1m` and `[0m` are incorrectly decoded console characters for `Console.BOLD` and `Console.RESET`, respectively.
I temporarily fixed it by removing those characters from the help message, but we should probably address this separately, later

Gedochao committed:
Adjust help-relevant tests

Gedochao committed:
Adjust `--help-js` and `--help-native`

Gedochao committed:
Apply review comments
- temporarily remove console-specific characters from help messages to fix the reference doc website
- remove redundant tags for options already tagged as `SpecificationLevel.MUST`
- rename `tags.important` to `tags.inShortHelp`

Gedochao commented:
Btw. as a side effect of these changes, I had to adjust the behaviour of `--help-js` and `--help-native` as well.
Previously, they simply ran the default `--help` with `Scala.js` or `Scala Native` options' sections included (alongside others).
Now, I instead run `--help-full` for just those help options groups.

old output:
```
▶ scala-cli --help-native
Usage: default [options]

Scala options:
  --js                                           Enable Scala.js. To show more options for Scala.js pass `--help-js`
  --native                                       Enable Scala Native. To show more options for Scala Native pass `--help-native`
  -P, --plugin, --compiler-plugin string*        Add compiler plugin dependencies
  -O, --scala-opt, --scala-option, --scalac-option option  Add a `scalac` option. Note that options starting with `-g`, `-language`, `-opt`, `-P`, `-target`, `-V`, `-W`, `-X`, and `-Y` are assumed to be Scala compiler options and don't require to be passed after `-O` or `--scalac-option`.
  -S, --scala, --scala-version version           Set the Scala version (3.2.2 by default)
  --scalac-help, --help-scalac                   Show help for scalac. This is an alias for --scalac-option -help
  --scalac-verbose, --verbose-scalac             Turn verbosity on for scalac. This is an alias for --scalac-option -verbose
  --script-snippet string*                       Allows to execute a passed string as a Scala script
  --scala-snippet string*                        Allows to execute a passed string as Scala code
  -with-compiler, --with-compiler, --with-scala-compiler  Allows to include the Scala compiler artifacts on the classpath.

Java options:
  --java-home path                               Set the Java home directory
  -j, --jvm jvm-name                             Use a specific JVM, such as `14`, `adopt:11`, or `graalvm:21`, or `system`
  --java-snippet string*                         Allows to execute a passed string as Java code
  --execute-java string*                         A synonym to --scala-snippet, which defaults the sub-command to `run` when no sub-command is passed explicitly
  -cp, --jar, --jars, --class, --classes, -classpath, --classpath, --extra-jar, --extra-jars, --class-path, --extra-class, --extra-classes, --extra-class-path paths  Add extra JARs and compiled classes to the class path
  --compile-only-jar, --compile-only-jars, --extra-compile-only-jar, --extra-compile-only-jars paths  Add extra JARs in the compilaion class path. Mainly using to run code in managed environments like Spark not to include certain depenencies on runtime ClassPath.
  --source-jar, --source-jars, --extra-source-jar, --extra-source-jars paths  Add extra source JARs
  --resource-dir, --resource-dirs paths          Add a resource directory
  -J, --java-opt java-options                    Set Java options, such as `-Xmx1g`
  --java-prop, --java-prop-option key=value|key  Add java properties. Note that options equal `-Dproperty=value` are assumed to be java properties and don't require to be passed after `--java-prop`.
  -J, --java-opt java-options                    Set Java options, such as `-Xmx1g`
  --java-prop, --java-prop-option key=value|key  Add java properties. Note that options equal `-Dproperty=value` are assumed to be java properties and don't require to be passed after `--java-prop`.

Repl options:
  -A, --amm, --ammonite                       Use Ammonite (instead of the default Scala REPL)
  --ammonite-ver, --ammonite-version string?  Set the Ammonite version (2.5.6-1-f8bff243 by default)

Logging options:
  -v, -verbose, --verbose  Increase verbosity (can be specified multiple times)
  -i, --interactive        Interactive mode
  --actions                Enable actionable diagnostics
  -q, --quiet              Decrease verbosity
  --progress               Use progress bars

Entrypoint options:
  -M, --main-class main-class                    Specify which main class to run
  --main-class-ls, --list-main-class, --main-class-list, --list-main-classes  List main classes available in the current context

Debug options:
  --debug                         Turn debugging on
  --debug-port string?            Debug port (5005 by default)
  --debug-mode attach|a|listen|l  Debug mode (attach by default)

Compilation server options:
  --server  Enable / disable usage of Bloop compilation server. Bloop is used by default so use `--server=false` to disable it. Disabling compilation server allows to test compilation in more controlled mannter (no caching or incremental compiler) but has a detrimental effect of performance.

Run options:
  --standalone-spark, --spark-standalone  Run as a Spark job, using a vanilla Spark distribution downloaded by Scala CLI
  --hadoop, --hadoop-jar                  Run as a Hadoop job, using the "hadoop jar" command
  --command                               Print the command that would have been run (one argument per line), rather than running it
  --scratch-dir string?                   Temporary / working directory where to write generated launchers

Benchmarking options:
  --jmh                  Run JMH benchmarks
  --jmh-version version  Set JMH version

Markdown options:
  --md-snippet, --markdown-snippet string*  [experimental] Allows to execute a passed string as Markdown code
  --md, --markdown, --enable-markdown       Enable markdown support.

Scala Native options:
  --native-version string?  Set the Scala Native version (0.4.9 by default).
  --native-mode string?     Set Scala Native compilation mode
  --native-gc string?       Set the Scala Native garbage collector
  --native-clang string?    Path to the Clang command
  --native-clangpp string?  Path to the Clang++ command
  --native-linking string*  Extra options passed to `clang` verbatim during linking
  --native-compile string*  List of compile options
  --embed-resources         Embed resources into the Scala Native binary (can be read with the Java resources API)

Dependency options:
  --dep, --dependency string*       Add dependencies
  -r, --repo, --repository string*  Add repositories

Other options:
  --suppress-directives-in-multiple-files-warning, --suppress-warning-directives-in-multiple-files  Suppress warnings about using directives in multiple files
  --home, --home-directory string?               Override the path to user's home directory
  --python-setup                                 Set Java options so that Python can be loaded
  --py, --python                                 Enable Python support via ScalaPy
  --scalapy-version, --scala-py-version string?  [experimental] Set ScalaPy version (0.5.3 by default)
  --platform scala-js|scala-native|jvm           Specify platform
  --help-js                                      Show options for ScalaJS
  --help-native                                  Show options for ScalaNative
  --help-doc, --doc-help, --help-scaladoc, --scaladoc-help  Show options for Scaladoc
  --help-repl, --repl-help                       Show options for Scala REPL
  --help-fmt, --fmt-help, --help-scalafmt, --scalafmt-help  Show options for Scalafmt
  -d, --compile-out, --destination, --compile-output, --output-directory, --compilation-output /example/path  Copy compilation results to output directory using either relative or absolute path
  --toolkit, --with-toolkit version|latest       Add toolkit to classPath
  -w, --watch                                    Watch source files for changes
  --restart, --revolver                          Run your application in background and automatically restart if sources have been changed
  --cross                                        Run given command against all provided Scala versions and/or platforms
  -w, --watch                                    Watch source files for changes
  --restart, --revolver                          Run your application in background and automatically restart if sources have been changed
  --cross                                        Run given command against all provided Scala versions and/or platforms
  -version, --version
```

new output:
```
Usage: default [options]

Scala Native options:
  --native-version string?   Set the Scala Native version (0.4.9 by default).
  --native-mode string?      Set Scala Native compilation mode
  --native-gc string?        Set the Scala Native garbage collector
  --native-clang string?     Path to the Clang command
  --native-clangpp string?   Path to the Clang++ command
  --native-linking string*   Extra options passed to `clang` verbatim during linking
  --native-linking-defaults  (hidden) Use default linking settings
  --native-compile string*   List of compile options
  --native-compile-defaults  (hidden) Use default compile options
  --embed-resources          Embed resources into the Scala Native binary (can be read with the Java resources API)

Other options:
  --suppress-directives-in-multiple-files-warning, --suppress-warning-directives-in-multiple-files  Suppress warnings about using directives in multiple files
  --suppress-outdated-dependency-warning         Suppress warnings about outdated dependencies in project
  --workspace path                               (hidden) Directory where .scala-build is written
  --runner                                       (hidden) Should include Scala CLI runner on the runtime ClassPath. Runner is added by default for application running on JVM using standard Scala versions. Runner is used to make stack traces more readable in case of application failure.
  --semantic-db                                  (hidden) Generate SemanticDBs
  --add-stubs                                    (hidden) Add dependency for stubs needed to make $ivy and $dep imports to work.
  --default-forbidden-directories                (hidden)
  --forbid string*                               (hidden)
  --strict-bloop-json-check                      (hidden)
  -version, --version
```

This probably should be tweaked further separately at a later date. 
`case-app` currently doesn't seem to support filtering out `Other options`, for example

Gedochao committed:
Ensure all launcher options are properly tagged

lwronski reviewed (approved):
LGTM

MaciejG604 commented on modules/cli/src/main/scala/scala/cli/commands/export0/ExportOptions.scala:
```diff
@@ -7,55 +7,64 @@ import scala.cli.commands.shared.{HasSharedOptions, HelpMessages, MainClassOptio
     mill: Option[Boolean] = None,
   @Tag(tags.restricted)
   @HelpMessage("Sets the export format to Json")
     json: Option[Boolean] = None,
```
Probably `@Group("Build Tool export")` and `@Tag(tags.important)` here, right?

Gedochao replied:
Yep, good catch, was added in parallel and I'd have missed it after the rebase...

Gedochao replied:
Fixed

Gedochao committed:
Ensure all `export` options are properly tagged

tgodzik commented on modules/cli/src/main/scala/scala/cli/commands/export0/ExportOptions.scala:
```diff
@@ -7,55 +7,66 @@ import scala.cli.commands.shared.{HasSharedOptions, HelpMessages, MainClassOptio
     mainClass: MainClassOptions = MainClassOptions(),
 
-  @Group("Build Tool export options")
+  @Group("Build Tool export")
```
Would love to see enums here, but we would need to do toString on that, but might be better than using just string everywhere :thinking: ?

Could be a follow up for sure

Gedochao replied:
Yeah, I was thinking about adding enums, but definitely not within the scope of this PR

tgodzik reviewed (approved):
Great work! I left a very minor comment for a future refactor

Gedochao committed:
Fix integration tests after tied with `--power` mode after relevant tags changed
